--- a/arch/arm/mach-cns3xxx/Kconfig
+++ b/arch/arm/mach-cns3xxx/Kconfig
@@ -14,6 +14,14 @@
 
 if ARCH_CNS3XXX
 
+config MACH_CNS3XXX
+	bool "Support CNS3XXX Evaluation Board 1.x"
+	help
+	  Include support for the Cavium Networks CNS3XXX MPCore Platform Baseboard.
+	  This is a platform with an on-board ARM11 MPCore and has support for USB, 
+	  USB-OTG, MMC/SD/SDIO and PCI-E, etc. It is used in several routers like
+	  the DLink DSR-250N and Netgear FVS318N.
+
 config MACH_CNS3420VB
 	bool "Support for CNS3420 Validation Board"
 	depends on ATAGS
--- a/arch/arm/mach-cns3xxx/Makefile
+++ b/arch/arm/mach-cns3xxx/Makefile
@@ -8,5 +8,6 @@
 cns3xxx-$(CONFIG_MACH_CNS3420VB)	+= cns3420vb.o
 cns3xxx-$(CONFIG_SMP)			+= platsmp.o headsmp.o cns3xxx_fiq.o
 cns3xxx-$(CONFIG_HOTPLUG_CPU)		+= hotplug.o
+cns3xxx-$(CONFIG_MACH_CNS3XXX)		+= cns3xxxevb.o
 cns3xxx-$(CONFIG_MACH_GW2388)		+= laguna.o
 
--- a/arch/arm/mach-cns3xxx/devices.c
+++ b/arch/arm/mach-cns3xxx/devices.c
@@ -361,6 +361,26 @@
 }
 
 /*
+ * I2S
+ */
+static struct resource cns3xxx_i2s_resource = {
+	.start		= CNS3XXX_I2S_BASE,
+	.end		= CNS3XXX_I2S_BASE + 0x53,
+	.flags		= IORESOURCE_MEM,
+};
+
+static struct platform_device cns3xxx_i2s_pdev = {
+	.name		= "cns3xxx-i2s",
+	.num_resources	= 1,
+	.resource	= &cns3xxx_i2s_resource,
+};
+
+void __init cns3xxx_i2s_init(void)
+{
+	platform_device_register(&cns3xxx_i2s_pdev);
+}
+
+/*
  * SPI
  */
 static struct resource cns3xxx_spi_resource = {
--- a/arch/arm/mach-cns3xxx/devices.h
+++ b/arch/arm/mach-cns3xxx/devices.h
@@ -22,6 +22,7 @@
 void __init cns3xxx_usb_init(void);
 void __init cns3xxx_usb_otg_init(void);
 void __init cns3xxx_i2c_init(void);
+void __init cns3xxx_i2s_init(void);
 void __init cns3xxx_spi_init(void);
 void __init cns3xxx_watchdog_init(void);
 
